File: sdk/py/hop_top_kit/safety.py
"""safety — delegation safety guard (12-factor AI CLI, Factor 10).

Prevents dangerous operations without explicit --force. In non-TTY
environments, caution-level ops also require --force.
"""

from __future__ import annotations

import sys
from enum import Enum


class SafetyLevel(Enum):
    READ = "read"
    CAUTION = "caution"
    DANGEROUS = "dangerous"


def safety_guard(level: SafetyLevel, *, force: bool = False) -> None:
    """Raise SystemExit if operation not permitted at given safety level."""
    if level == SafetyLevel.READ:
        return

    if level == SafetyLevel.DANGEROUS and not force:
        raise SystemExit(f"Refusing {level.value} operation without --force")

    if level == SafetyLevel.CAUTION and not force and not sys.stdin.isatty():
        raise SystemExit(f"Refusing {level.value} operation in non-TTY without --force")

File: sdk/ts/src/safety.ts
/**
 * @module safety
 * @package @hop-top/kit
 *
 * Factor 10 — Delegation Safety.
 *
 * Guards destructive operations behind --force, with stricter
 * requirements in non-TTY (scripted/agent) contexts.
 */

export type SafetyLevel = 'read' | 'caution' | 'dangerous';

/**
 * Validates that the current context permits the requested safety level.
 *
 * - `read`: always passes
 * - `caution`: passes in TTY; requires `--force` in non-TTY
 * - `dangerous`: always requires `--force`
 *
 * @throws {Error} when guard fails, with message referencing --force
 */
export function safetyGuard(
  level: SafetyLevel,
  opts?: { force?: boolean },
): void {
  if (level === 'read') return;

  const force = opts?.force ?? false;
  const isTTY = !!process.stdout.isTTY;

  if (level === 'dangerous' && !force) {
    throw new Error(
      'dangerous operation requires --force',
    );
  }

  if (level === 'caution' && !isTTY && !force) {
    throw new Error(
      'caution-level operation in non-TTY requires --force',
    );
  }
}
